src/to_vec.jl

```diff
@@ -53,7 +53,7 @@ function to_vec(x::StridedVector)
     return x_vec, Vector_from_vec
 end
 
-function to_vec(x::Union{SubArray, StridedArray})
+function to_vec(x::Union{SubArray, Base.ReshapedArray, StridedArray})
```

Comment:
Maybe the parent array of `SubArray` should be restricted to e.g. `StridedArray` or something?

Reply:
We want e.g. a
```julia
julia> sa = @view randn(T, 10)[1:4]
julia> typeof(sa)
SubArray{Float64, 1, Vector{Float64}, Tuple{UnitRange{Int64}}, true}
```
to dispatch on the current `SubArray` implementation.  Unfortunately `StridedArray` is a `Union` of several more specific `SubArrays` (not just in the parent argument)

So we have two options:
1) Narrow the dispatch of the current `SubArray` (very verbose*)
2) Loosen the dispatch for `StridedArray` to `Union{StridedArray, SubArray}` so that the current `SubArray` becomes more specific.

And similarly for the `ReshapedArray`.

I don't particularly like either option but it's probably better than the status quo.

*Number 1) looks like this, I want to nope out of that 😂 

```julia
StridedSubArray = Union{
SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real, N} where N}, Tuple{AbstractUnitRange, Vararg{Any, N} where N}}}, DenseArray}, MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}, N} where N}},
SubArray{T, 1, A, I, L} where {A<:Union{Base.ReinterpretArray{T, N, S, A, IsReshaped} where {T, N, A<:Union{SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real, N} where N}, Tuple{AbstractUnitRange, Vararg{Any, N} where N}}}, DenseArray}, IsReshaped, S},
SubArray{T, N, A, I, true} where {T, N, A<:DenseArray, I<:Union{Tuple{Vararg{Real, N} where N}, Tuple{AbstractUnitRange, Vararg{Any, N} where N}}}, DenseArray}, MI<:Tuple{Vararg{Base.MultiplicativeInverses.SignedMultiplicativeInverse{Int64}, N} where N}}, DenseArray}, I<:Tuple{Vararg{Union{Int64, AbstractRange{Int64}, Base.AbstractCartesianIndex, Base.ReshapedArray{T, N, A, Tuple{}} where {T, N, A<:AbstractUnitRange}}, N} where N}, L}} where T)
}
```

Reply:
@willtebbutt how do you feel about that?

Reply:
The thing I'm concerned about is if someone attempts to `to_vec` something like
```julia
view(Diagonal(randn(5)), 1:5, 1:3)
```

> *Number 1) looks like this, I want to nope out of that 😂

Lol yeah, that's really not fun.

Is there a reason that we can't go with something like
```julia
Union{StridedArray, SubArray{T, D, <:StridedArray}}
```
and just recurse into `ReshapedArray`s?

Reply:
The dispatch for the `Diagonal` example is the same (the dedicated `SubArray` method) whether we constrain the parent of the `SubArray` to be strided or not.

The example breaks, but because `SparseArray.SparseMatrixCSC` does not have a `to_vec` defined.

Reply:
Wait, we are able to `to_vec` a sparse array? That makes no sense to me.

Reply:
No no, sorry, I must have confused you. This is what happens with the `Diagonal` example:
```julia
julia> vd = view(Diagonal(randn(5)), 1:5, 1:3)
5×3 view(::Diagonal{Float64, Vector{Float64}}, 1:5, 1:3) with eltype Float64:
 0.168618   0.0        0.0
 0.0       -0.796772   0.0
 0.0        0.0       -1.16312
 0.0        0.0        0.0
 0.0        0.0        0.0

julia> v, back = to_vec(vd)
([5.0, 3.0, 1.0, 2.0, 3.0, 4.0, 1.0, 2.0, 3.0, 0.16861792127500824, -0.796772203387132, -1.1631233934390923], FiniteDifferences.var"#structtype_from_vec#29"{SparseMatrixCSC{Float64, Int64}, FiniteDifferences.var"#Tuple_from_vec#45"{NTuple{5, Int64}, NTuple{5, Int64}, NTuple{5, typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#22", FiniteDifferences.var"#Real_from_vec#22", typeof(identity), typeof(identity), typeof(identity)}}(FiniteDifferences.var"#Tuple_from_vec#45"{NTuple{5, Int64}, NTuple{5, Int64}, NTuple{5, typeof(identity)}}((1, 2, 6, 9, 12), (1, 1, 4, 3, 3), (identity, identity, identity, identity, identity)), (FiniteDifferences.var"#Real_from_vec#22"(), FiniteDifferences.var"#Real_from_vec#22"(), identity, identity, identity)))

julia> back(v)
ERROR: MethodError: no method matching SparseMatrixCSC{Float64, Int64}(::Float64, ::Float64, ::Vector{Float64}, ::Vector{Float64}, ::Vector{Float64})
Closest candidates are:
  SparseMatrixCSC{Tv, Ti}(::Integer, ::Integer, ::Vector{Ti}, ::Vector{Ti}, ::Vector{Tv}) where {Tv, Ti<:Integer} at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/SparseArrays/src/sparsematrix.jl:26
Stacktrace:
 [1] (::FiniteDifferences.var"#structtype_from_vec#29"{SparseMatrixCSC{Float64, Int64}, FiniteDifferences.var"#Tuple_from_vec#45"{NTuple{5, Int64}, NTuple{5, Int64}, NTuple{5, typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#22", FiniteDifferences.var"#Real_from_vec#22", typeof(identity), typeof(identity), typeof(identity)}})(v::Vector{Float64})
   @ FiniteDifferences ~/JuliaEnvs/ChainRules.jl/dev/FiniteDifferences/src/to_vec.jl:38
 [2] top-level scope
   @ REPL[85]:1
```

The dispatch (and the error message) is the same in both cases, whereas we have
```julia
function to_vec(x::Union{SubArray{T, D, <:StridedArray}, Base.ReshapedArray, StridedArray}) where {T, D}
```
or
```julia
function to_vec(x::Union{SubArray, Base.ReshapedArray, StridedArray})
```

Reply:
Lol -- I wonder how we wound up with that sparse array...

Reply:
> ([5.0, 3.0, 1.0, 2.0, 3.0, 4.0, 1.0, 2.0, 3.0, 0.16861792127500824, -0.796772203387132, -1.1631233934390923], FiniteDifferences.var"#structtype_from_vec#29"{SparseMatrixCSC{Float64, Int64}, FiniteDifferences.var"#Tuple_from_vec#45"{NTuple{5, Int64}, NTuple{5, Int64}, NTuple{5, typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#22", FiniteDifferences.var"#Real_from_vec#22", typeof(identity), typeof(identity), typeof(identity)}}(FiniteDifferences.var"#Tuple_from_vec#45"{NTuple{5, Int64}, NTuple{5, Int64}, NTuple{5, typeof(identity)}}((1, 2, 6, 9, 12), (1, 1, 4, 3, 3), (identity, identity, identity, identity, identity)), (FiniteDifferences.var"#Real_from_vec#22"(), FiniteDifferences.var"#Real_from_vec#22"(), identity, identity, identity)))

And where does the `[5.0, 3.0, 1.0...]` come from 🤔

Reply:
Oh, it's the method on line 81
```julia
to_vec(x::Base.SubArray) = to_vec(copy(x))
```

Reply:
Yeah, if you remove that method and the `SubArray`s from the unions, it seems to work, which makes sense to me.

Reply:
So what's the upshot? Keep as is or solve the `SparseMatrixCSC` issue as well? I feel like that is a separate PR?

Reply:
I'm confused. My understanding of the situation is
1. the SparseMatrixCSC stuff was an artifact of something going wrong
2. removing `to_vec` for `Base.SubArray`, and removing `SubArray` from the Union makes everything work.
Is this not the case?

Reply:
The status is:
- All tests pass with the current PR
- Removing the `to_vec(Base.SubArray)` breaks many tests
- `test_to_vec(view(Diagonal(randn(5)), 1:5, 1:3); check_inferred=false)`, (the thing you suggested should be checked) breaks on master (`StackOverflowError`), on the current PR (`SparseMatrixCSC` error), and on the current PR with `SubArray` removed from the Union (`SparseMatrixCSC` error).

Reply:
Ahh I see.

I've run it further locally, and found that the following seems to resolve everything for me:

1. restrict the array methods to `DenseVector` and `DenseArray`, from `Union{StridedArray...}`
2. add a specific method for `SubArray`:
```julia
function to_vec(x::SubArray)
    x_vec, from_vec = to_vec(x.parent)
    SubArray_from_vec(x_vec) = view(from_vec(x_vec), x.indices...)
    return x_vec, SubArray_from_vec
end
```
if you don't have this, it thinks that the integer fields should also be `to_vec`ed.
3. comment out the `SubArray`-specific method

Reply:
Oh yeah that's much better. Are we happy with that?

Reply:
I'm happy with it, but I feel that we should get an additional review. Maybe @oxinabox or @sethaxen ?

Reply:
Actually with this one ChainRules test fails, and the reason is that the current implementation returns the whole parent vector
```julia
julia> a = rand(7);

julia> b = @view a[2:5];

julia> v, back = to_vec(b);

julia> v
7-element Vector{Float64}:
 0.9586867155061674
 0.8025760289977295
 0.9784238101274141
 0.16440182965461236
 0.17784328126321847
 0.01704930118685577
 0.8408143603848348
```

Keeping the old version below works just fine. No issues now `StridedArray` has been replaced by `DenseArray` (where we used to have `AbstractArray`)
```julia
to_vec(x::SubArray) = to_vec(copy(x))
```